src/pal/src/loader/module.cpp

```diff
@@ -1339,29 +1339,33 @@ static HMODULE LOADLoadLibrary(LPCSTR shortAsciiName, BOOL fDynamic)
     // See if file can be dlopen()ed; this should work even if it's already loaded
     {
         // See GetProcAddress for an explanation why we leave the PAL.
-        PAL_LeaveHolder holder;
-        dl_handle = dlopen(shortAsciiName, RTLD_LAZY | RTLD_NOLOAD); 
-        if (!dl_handle)
+        PAL_LeaveHolder holder; 
+        
+        // P/Invokes are often declared with variations on the actual library name.
+        // For example, it's common to leave off the extension/suffix of the library
+        // even if it has one, or to leave off a prefix like "lib" even if it has one
+        // (both of these are done typically to smooth over cross-platform differences). 
+        // We try to dlopen with such variations on the original.
+        const char* const formatStrings[4] = // used with args: PAL_SHLIB_PREFIX, shortAsciiName, PAL_SHLIB_SUFFIX
         {
-            dl_handle = dlopen(shortAsciiName, RTLD_LAZY);
-        }
-
-        // P/Invoke calls are often defined without an extension in the name of the 
-        // target library. So if we failed to load the specified library, try adding
-        // a proper extension and load the library again.
-        if (!dl_handle)
+            "%s%s%s",     // prefix+name+suffix
+            "%.0s%s%.0s", // name
```

Comment:
Oh, I missed it. The entire conversion was hidden under “outdated diff” and I did not notice it. I’m sorry about that.

My suggestion was based on a thought that first of all we should try to load a library exactly as the caller specified without trying to guess what the caller meant to load and only if that fails we should try to play the game of guessing. Basically if we take your example with “CoreFoundation” (which does not have a prefix) and we first try to load “libcorefoundation.dylib” then we potentially can load wrong library and there will be no way for a developer to workaround that. So, personally, I would prefer the original order that you suggested: name, namesuffix, prefixnamesuffix, and probably prefixname.

Of course, if everyone else thinks that the current order make more sense then I’m OK with it.